janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphLocalQueryOptimizerStrategy.java

```diff
@@ -56,8 +56,16 @@ public void apply(final Traversal.Admin<?, ?> traversal) {
 
         final Graph graph = traversal.getGraph().get();
 
+        StandardJanusGraph janusGraph;
+        if (graph instanceof StandardJanusGraph) {
+            janusGraph = (StandardJanusGraph) graph;
+        } else if (graph instanceof StandardJanusGraphTx) {
+            janusGraph = ((StandardJanusGraphTx) graph).getGraph();
+        } else {
+            return;
```

Comment:
This is required because now graph can be an [EmptyGraph](http://tinkerpop.apache.org/javadocs/3.5.0/full/org/apache/tinkerpop/gremlin/structure/util/empty/EmptyGraph.html) instance. Seems to be related to https://github.com/apache/tinkerpop/pull/1211/files

Due to the changes in traversal strategy methodology,  now some traversal relevant tests are failing. For example, JanusGraphStepStrategyTest::doMultiQueryTest.

```
Expected :[JanusGraphStep(vertex,[]), JanusGraphMultiQueryStep(ChooseStep), ChooseStep([JanusGraphVertexStep([weight.eq(0)]), HasNextStep],[[(eq(true)), [JanusGraphVertexStep([weight.eq(1)]), EndStep]], [(eq(false)), [IdentityStep, EndStep]]])]
Actual   :[JanusGraphStep(vertex,[]), JanusGraphMultiQueryStep(ChooseStep), ChooseStep([VertexStep(IN,[knows],edge), HasStep([weight.eq(0)]), HasNextStep],[[(eq(true)), [VertexStep(IN,[knows],edge), HasStep([weight.eq(1)]), EndStep]], [(eq(false)), [IdentityStep, EndStep]]])]
```

Reply:
So to conclude, the problem seems to be that child traversals use `EmptyGraph` for some reason instead of the parent traversal's graph? At least that's what I read from your comment on the TinkerPop PR.

Reply:
@rngcntr Yes.

Reply:
I agree that this looks very much like a problem with TinkerPop. As a workaround, we could obviously catch these cases and set the graph to match the parent graph, manually. But I hope that TP recognizes the issue and hopefully fixes it with 3.5.1 soon. To raise awareness of this bug, I created a [bug ticket](https://issues.apache.org/jira/projects/TINKERPOP/issues/TINKERPOP-2567).

Reply:
> As a workaround, we could obviously catch these cases and set the graph to match the parent graph

I am curious about how to do that.

Reply:
My intuition would be to create a strategy that runs prior to all of our other strategies and does nothing but setting the graph of each traversal to their parent's graph. Once the problem is fixed on the TP side, we could just drop that strategy.

Reply:
Possible. Alternatively, we could do

```java
graph = traversal.getParent().asStep().getTraversal().getGraph().get();
```

recursively until we find StandardJanusGraph.

Reply:
Let's wait for the response from your TinkerPop JIRA ticket. I prefer delaying this upgrade till possibly the next JanusGraph release.

Reply:
Just as a side note, I really like the new execution order of strategies in general. The old execution order has caused lots of pain in some of my PRs and the new one makes many things easier!

Reply:
@rngcntr that's good to know. i'm glad i came back to this issue to see that comment.

Reply:
@spmallette Being able to rely on the fact that all strategies from `applyPrior()` are already applied to child traversals really makes a difference.